### ISSUE TITLE
`reduce` and `foldr`/`foldl` are inconsistent for + and empty vectors

### ISSUE THREAD

**Issue by dlfivefifty** [JuliaLang/julia#20144]
In 0.6, `reduce` and `foldr`/`foldl` are inconsistent for empty vectors:
```julia
julia> reduce(+,Float64[])
0.0

julia> foldr(+,Float64[])
ERROR: BoundsError: attempt to access 0-element Array{Float64,1} at index [0]
Stacktrace:
 [1] foldr(::Function, ::Array{Float64,1}) at ./reduce.jl:160

julia> foldl(+,Float64[])
0.0
```

**Comment by dlfivefifty** [comment]
Related issue is https://github.com/JuliaLang/julia/issues/5797#issuecomment-273974115

**Comment by vsart** [comment]
I'm new to Julia so take all of this with a grain of salt.

So the `foldr` with two arguments ([here](https://github.com/JuliaLang/julia/blob/master/base/reduce.jl#L160)) is calling `mapfoldr` with three arguments ([here](https://github.com/JuliaLang/julia/blob/master/base/reduce.jl#L134)).

It's defined as 
`mapfoldr(f, op, itr) = (i = endof(itr); mapfoldr_impl(f, op, f(itr[i]), itr, i-1))`

So not only would it send `-1` as the `i` argument to `mapfoldr_impl`, it also tries to apply the function to `itr[i]`, which is the 0-index access that happened here.

I think the check for `i == 0` in `mapfoldr_impl` should instead happen before the call, since to pass the `v0` argument, it needs to access a member of the array (which in this case is impossible).

`reduce` and `foldl` seem to use `mapfoldl`, which does the proposed check ([here](https://github.com/JuliaLang/julia/blob/master/base/reduce.jl#L68)).

Should be an easy fix if I'm right.

**Comment by vsart** [comment]
So, the check on `foldl`/`reduce` was added as a fix for #7465, but it wasn't added to `foldr`, so I think it really is intended for a check to exist in `mapfoldr` for empty arguments.

I'm gonna go ahead and work on a PR if that's okay.

**Comment by vsart** [comment]
This issue should be closed (see #20160).

**Comment by Sacha0** [comment]
Thanks @vsart!
